web/controllers/userController/index.ts

```diff
@@ -1007,7 +1011,66 @@ async function putActive(wix_id: string, active: boolean, person: Pupil | Studen
                 await dissolveMatch(matches[i], 0, person);
             }
 
-            // Step 2: Deactivate
+            // Step 2: Cancel all courses if user is student
+            if (type == Student) {
+                let courses = await getConnection()
+                    .getRepository(Course)
+                    .createQueryBuilder("course")
+                    .leftJoinAndSelect("course.instructors", "instructors")
+                    .getMany();
+
+                courses.forEach(async (course: Course) => {
+                    if (course.instructors.some(i => i.id === person.id)) { // Only proceed if we're part of this course as an instructor
+                        if (course.instructors.length > 1) {
+                            // Course still has other instructors, only remove our person from those. We don't want to cancel those courses.
+                            course.instructors = course.instructors.filter(s => s.id !== person.id);
+                            entityManager.transaction(async em => {
+                                await em.save(Course, course);
+                                logger.info("Removed instructor " + person.firstname + " " + person.lastname + " from course " + course.name + ".");
+                            });
+                        } else {
+                        // Our person is the only instructor in the course. Cancel it.
+
+                            // We have a non-mitigated race condition here: Someone could post a new subcourse into the course, while the course gets cancelled
+                            try {
+                            // Run in transaction, so we may not have a mixed state, where some subcourses are cancelled, but others are not
+                                await entityManager.transaction(async em => {
+                                    if (course.hasOwnProperty("subcourses")) {
+                                        for (let i = 0; i < course.subcourses.length; i++) {
+                                            if (!course.subcourses[i].cancelled) {
+                                                course.subcourses[i].cancelled = true;
+                                                await em.save(Subcourse, course.subcourses[i]);
+                                                sendSubcourseCancelNotifications(course, course.subcourses[i]);
+                                            }
+                                        }
+                                    }
+
+                                    course.courseState = CourseState.CANCELLED;
+                                    await em.save(Course, course);
+
+                                }).catch(e => {
```

Comment:
Can we really handle the error at this point? Through catching here the user gets deactivated, but some courses are in an inconsistent state (I know that we have similar patterns at other places though). I'd personally prefer to "fail and rollback" instead of "half deleting" the user.

Reply:
I see what you mean. How would the rollback work? Is this handled automatically by the entityManager if it wraps the entire code? (see latest commit).
And yeah, this code block was largely copied from another endpoint, and I didn't pay much attention to those things you've commented on. I shouldn't have created a PR, since it wasn't production ready :(

Reply:
@Jonasdoubleyou

Reply:
Yes, it would if you let all errors bubble up to the entity manager callback (which catches the error, aborts the transaction, and then rethrows it). So just remove the inner `try { ... } catch(e) { .. }` and let the error bubble up.

Reply:
The TypeORM documentation is lacking as always, you can find the implementation behind `.transaction(...)` [here](https://github.com/typeorm/typeorm/blob/e9366b33ddff296de1254019589b85e40aa53e12/src/entity-manager/EntityManager.ts#L145)